pkg/adapter/awssnssource/reconciler.go

```diff
@@ -0,0 +1,112 @@
+Copyright (c) 2021 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awssnssource
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/reconciler"
+
+	"github.com/triggermesh/aws-event-sources/pkg/adapter/awssnssource/status"
+	"github.com/triggermesh/aws-event-sources/pkg/apis/sources/v1alpha1"
+	reconcilerv1alpha1 "github.com/triggermesh/aws-event-sources/pkg/client/generated/injection/reconciler/sources/v1alpha1/awssnssource"
+)
+
+// Reconciler implements controller.Reconciler for the event source type.
+type Reconciler struct {
+	adapter MTAdapter
+}
+
+// Check the interfaces Reconciler should implement.
+var (
+	_ reconcilerv1alpha1.Interface         = (*Reconciler)(nil)
+	_ reconcilerv1alpha1.ReadOnlyInterface = (*Reconciler)(nil)
+	_ reconcilerv1alpha1.Finalizer         = (*Reconciler)(nil)
+	_ reconcilerv1alpha1.ReadOnlyFinalizer = (*Reconciler)(nil)
+)
+
+// ReconcileKind implements reconcilerv1alpha1.Interface.
+func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.AWSSNSSource) reconciler.Event {
+	if err := r.reconcile(ctx, src); err != nil {
+		return err
+	}
+
+	// TODO(antoineco): switch to true and adjust reason and message once
+	// routing is implemented
+	cond := &apis.Condition{
+		Type:    status.ConditionHandlerRegistered,
+		Status:  corev1.ConditionFalse,
+		Reason:  "TODO",
+		Message: "Multi-tenancy not implemented",
+	}
+
+	if err := r.adapter.PropagateCondition(ctx, src, cond); err != nil {
+		return fmt.Errorf("propagating status condition: %w", err)
+	}
+
+	return nil
+}
+
+// ObserveKind implements reconcilerv1alpha1.ReadOnlyInterface.
+func (r *Reconciler) ObserveKind(ctx context.Context, src *v1alpha1.AWSSNSSource) reconciler.Event {
+	return r.reconcile(ctx, src)
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, src *v1alpha1.AWSSNSSource) error {
+	if !src.Status.IsReady() {
+		// Mark that error as permanent so we don't retry until the
+		// source's status has been updated, which automatically
+		// triggers a new reconciliation.
+		return controller.NewPermanentError(reconciler.NewEvent(corev1.EventTypeWarning, "NotReady",
+			"Source is not ready yet. Skipping adapter configuration"))
+	}
+
+	err := r.adapter.RegisterHandlerFor(ctx, src)
+	if err != nil {
+		return reconciler.NewEvent(corev1.EventTypeWarning, ReasonHandlerFailedRegister,
+			"Failed to register HTTP handler: %s", err)
+	}
```

Comment:
Here is an example of call from the reconciler to a method exposed by the adapter.